### PR TITLE
feat: show accurate expiration time for swap rates

### PR DESCRIPTION
**Pull request by cjordan-ledger**
### 📝 Description

When a user is looking to swap currencies, we will check if any of the returned exchange rates expire at a specific time, and if so, we then pick the closest expirationTime and set our refresh/countdown logic to use this value. If none of the exchange rates have an expirationTime value, we default to 60s

Note: we still only show the countdown/refresh the rates if there are no errors on this screen.

### ❓ Context

- **Impacted projects**:  LLD
- **Linked resource(s)**:  https://ledgerhq.atlassian.net/browse/LIVE-4682

### ✅ Checklist

- [ ] **Test coverage** 
- [ ] **Atomic delivery** 
- [ ] **No breaking changes** 

### Showing refresh time based on fixed exchange rate
(note: we won't actually show the countdown when there's an error. I've removed some logic that checks for this whilst taking this screenshot to demo the new countdown duration is working correctly )
image

### 🚀 Expectations to reach

_Please make sure you follow these [**Important Steps**](https://github.com/LedgerHQ/ledger-live/blob/develop/CONTRIBUTING.md#important-steps)._

_Pull Requests must pass the CI and be internally validated in order to be merged._

### CODE REVIEW

File: apps/ledger-live-desktop/src/renderer/screens/exchange/Swap2/Form/hooks/useRefreshRates.js [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+  const refreshTime = useMemo(() => getRefreshTime(swap.rates?.value), [swap.rates?.value]);
+
+  useEffect(() => {
+    refreshInterval.current && clearInterval(refreshInterval.current);
+    refreshInterval.current = setInterval(() => {
+      if (!pause) {
+        swap.refetchRates();
+      }
+    }, refreshTime);
+  }, [refreshTime, pause, swap, swap.rates.value]);
```
Comment (sarneijim) [inline apps/ledger-live-desktop/src/renderer/screens/exchange/Swap2/Form/hooks/useRefreshRates.js:19]: }, [refreshTime, pause, swap.rates?.value]);

Comment (cjordan-ledger) [reply apps/ledger-live-desktop/src/renderer/screens/exchange/Swap2/Form/hooks/useRefreshRates.js]: https://github.com/LedgerHQ/ledger-live/pull/2313/commits/e9a87e8218e0a4812bd2e2eaf9eab0dd02a13dce.

**Comment by cjordan-ledger** [comment]
> Funny behaviour here with 65 seconds showing:
> 
>  65.seconds.mov

@ggilchrist-ledger just to confirm we are showing 65 seconds here as that is how much time before the rate expires based on the lowest `expirationTime` captured from `/rates`.

**Comment by sarneijim** [comment]
> > Funny behaviour here with 65 seconds showing:
> > 65.seconds.mov
> 
> @ggilchrist-ledger just to confirm we are showing 65 seconds here as that is how much time before the rate expires based on the lowest `expirationTime` captured from `/rates`.

@jnicoulaud-ledger , are you aware of this?

**Comment by ggilchrist-ledger** [comment]
@sarneijim I let the 60 second timer expire and the `Loading...` text was there for about 30 seconds. Probably not network slowness as generating new quotes is very quick 
image

**Comment by ggilchrist-ledger** [comment]
https://ledgerhq.atlassian.net/browse/LIVE-4682?focusedCommentId=272800

**Comment by sarneijim** [comment]
> @sarneijim I let the 60 second timer expire and the `Loading...` text was there for about 30 seconds. Probably not network slowness as generating new quotes is very quick image

I have been retesting. I do not see any issue.
I assume it was a network error, as took 30s, not 60s.

**Comment by cjordan-ledger** [comment]
@ggilchrist-ledger @sarneijim changes Some pushed since you last reviewed - we now format the countdown as mm:ss (i.e. show 1:00 instead of 0:60) - and with the refresh duration we either show 60s or the expirationRate, and choose the minimum of these two values.

File: apps/ledger-live-desktop/src/renderer/screens/exchange/Swap2/Form/Rates/utils/formatCountdown.test.js [hunk 2]
```diff
@@ -0,0 +1,12 @@
+// @flow
+import { formatCountdown } from "~/renderer/screens/exchange/Swap2/Form/Rates/utils/formatCountdown";
+
+describe("formatCountdown", () => {
+  it("formats countdown to timer", () => {
+    expect(formatCountdown(1)).toBe("00:01");
+    expect(formatCountdown(10)).toBe("00:10");
+    expect(formatCountdown(60)).toBe("01:00");
+    expect(formatCountdown(61)).toBe("01:01");
+    expect(formatCountdown(-1)).toBe("00:01");
```
Comment (ggilchrist-ledger) [inline apps/ledger-live-desktop/src/renderer/screens/exchange/Swap2/Form/Rates/utils/formatCountdown.test.js:10]: When would this happen? @cjordan-ledger

Comment (cjordan-ledger) [reply apps/ledger-live-desktop/src/renderer/screens/exchange/Swap2/Form/Rates/utils/formatCountdown.test.js:10]: There is existing logic that makes negative values positive...

https://github.com/LedgerHQ/ledger-live/pull/2313/files#diff-3491da745355daff4fdfd891af2156dfd2fedcfb11db22e49de080e45bf34ac7L51

But looking at the code I think it'll actually be best to remove this as we already check if the countdown is positive before displaying it

Comment (cjordan-ledger) [reply apps/ledger-live-desktop/src/renderer/screens/exchange/Swap2/Form/Rates/utils/formatCountdown.test.js]: https://github.com/LedgerHQ/ledger-live/pull/2313/commits/fe514c60dd12e7e2b9a0f63794230e82c9153ad4

File: apps/ledger-live-desktop/src/renderer/screens/exchange/Swap2/Form/hooks/useRefreshRates.js [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+  const refreshInterval = useRef();
+
+  const refreshTime = useMemo(() => getRefreshTime(swap.rates?.value), [swap.rates?.value]);
+
+  useEffect(() => {
+    clearTimeout(refreshInterval.current);
+
+    refreshInterval.current = setTimeout(() => {
+      console.log("useRefreshRates, calling refetchRates", refreshTime);
+      console.log("useRefreshRates, pause = ", pause);
```
Comment (sarneijim) [inline apps/ledger-live-desktop/src/renderer/screens/exchange/Swap2/Form/hooks/useRefreshRates.js:17]: Can we remove those console.log?

**Review by sarneijim — APPROVED** [review]

File: apps/ledger-live-desktop/src/renderer/screens/exchange/Swap2/Form/hooks/useRefreshRates.js (hunk 3 above)
Comment (cjordan-ledger) [reply apps/ledger-live-desktop/src/renderer/screens/exchange/Swap2/Form/hooks/useRefreshRates.js:17]: Ah yes, thanks for reminding. Wanted to keep this in until Gregor/Xavier has confirmed it's now working - then will remove

**Pull request merged**